L1 dates

Some types have `PrimitiveType: "timestamp"` in the old spec that we used to render as a date.

```
aws-s3/lib/bucket.ts:2076:9 - error TS2322: Type 'Date | undefined' is not assignable to type 'string | undefined'.
  Type 'Date' is not assignable to type 'string'.

2076         expirationDate: rule.expirationDate,
             ~~~~~~~~~~~~~~

  aws-s3/lib/s3.generated.ts:1263:14
    1263     readonly expirationDate?: string;
                      ~~~~~~~~~~~~~~
    The expected type comes from property 'expirationDate' which is declared here on type 'RuleProperty'
aws-s3/lib/bucket.ts:2090:9 - error TS2322: Type '{ storageClass: string; transitionDate: Date | undefined; transitionInDays: number | undefined; }[] | undefined' is not assignable to type 'IResolvable | (IResolvable | TransitionProperty)[] | undefined'.
  Type '{ storageClass: string; transitionDate: Date | undefined; transitionInDays: number | undefined; }[]' is not assignable to type 'IResolvable | (IResolvable | TransitionProperty)[] | undefined'.
    Type '{ storageClass: string; transitionDate: Date | undefined; transitionInDays: number | undefined; }[]' is not assignable to type '(IResolvable | TransitionProperty)[]'.
      Type '{ storageClass: string; transitionDate: Date | undefined; transitionInDays: number | undefined; }' is not assignable to type 'IResolvable | TransitionProperty'.
        Type '{ storageClass: string; transitionDate: Date | undefined; transitionInDays: number | undefined; }' is not assignable to type 'TransitionProperty'.
          Types of property 'transitionDate' are incompatible.
            Type 'Date | undefined' is not assignable to type 'string | undefined'.

2090         transitions: mapOrUndefined(rule.transitions, t => ({
```

--- a/packages/@aws-cdk/cfn-resources/src/cli/cdk/type-converter.ts
+++ b/packages/@aws-cdk/cfn-resources/src/cli/cdk/type-converter.ts
@@ -56,6 +56,8 @@ export class TypeConverter {
         return Type.NUMBER;
       case 'boolean':
         return Type.BOOLEAN;
+      case 'date-time':
+        return Type.DATE_TIME;
       case 'array':
         return Type.arrayOf(this.typeFromSpecType(type.element));
       case 'map':

--- a/packages/@aws-cdk/cfn-resources/src/cli/prop-mapping.ts
+++ b/packages/@aws-cdk/cfn-resources/src/cli/prop-mapping.ts
@@ -99,7 +99,7 @@ export class PropMapping {
           parse: CDK_CORE.helpers.FromCloudFormation.getString,
           validate: CDK_CORE.validateString,
         };
-      case PrimitiveType.Date:
+      case PrimitiveType.DateTime:
         return {
           produce: CDK_CORE.dateToCloudFormation,
           parse: CDK_CORE.helpers.FromCloudFormation.getDate,

--- a/packages/@aws-cdk/service-spec-build/src/build-database.ts
+++ b/packages/@aws-cdk/service-spec-build/src/build-database.ts
@@ -34,7 +34,14 @@ export async function buildDatabase(options: BuildDatabaseOptions = {}) {
         db,
         resource,
         fails: warnings,
-        specResource: resourceSpec.ResourceTypes[resource.typeName],
+        resourceSpec: {
+          spec: resourceSpec.ResourceTypes[resource.typeName],
+          types: Object.fromEntries(
+            Object.entries(resourceSpec.PropertyTypes)
+              .filter(([typeName]) => typeName.startsWith(resource.typeName))
+              .map(([typeName, typeDef]) => [typeName.split('.').splice(1).join('.'), typeDef]),
+          ),
+        },
       });
       db.link('regionHasResource', region, res);
 

--- a/packages/@aws-cdk/service-spec-build/src/import-cloudformation-registry.ts
+++ b/packages/@aws-cdk/service-spec-build/src/import-cloudformation-registry.ts
@@ -32,7 +32,10 @@ export interface LoadCloudFormationRegistryResourceOptions {
   readonly db: SpecDatabase;
   readonly resource: CloudFormationRegistryResource;
   readonly fails: Failures;
-  readonly specResource?: resourcespec.ResourceType;
+  readonly resourceSpec?: {
+    spec?: resourcespec.ResourceType;
+    types?: Record<string, resourcespec.PropertyType>;
+  };
 }
 export interface LoadCloudFormationRegistryServiceFromResourceOptions {
   readonly db: SpecDatabase;
@@ -68,7 +71,7 @@ export function importCloudFormationRegistryResource(options: LoadCloudFormation
       properties: {},
     });
 
-    recurseProperties(resource, res.properties, resourceFailure);
+    recurseProperties(resource, res.properties, resourceFailure, options.resourceSpec?.spec);
     // Every property that's a "readonly" property, remove it again from the `properties` collection.
     for (const propPtr of findAttributes(resource)) {
       const attrName = simplePropNameFromJsonPtr(propPtr);
@@ -90,13 +93,25 @@ export function importCloudFormationRegistryResource(options: LoadCloudFormation
     return res;
   }
 
-  function recurseProperties(source: ImplicitJsonSchemaRecord, target: ResourceProperties, fail: Fail) {
+  function recurseProperties(
+    source: ImplicitJsonSchemaRecord,
+    target: ResourceProperties,
+    fail: Fail,
+    spec?: resourcespec.PropertyType,
+  ) {
     if (!source.properties) {
       throw new Error(`Not an object type with properties: ${JSON.stringify(source)}`);
     }
 
     for (const [name, property] of Object.entries(source.properties)) {
-      const resolved = resolve(property);
+      let resolved = resolve(property);
+
+      if (spec?.Properties?.[name]?.PrimitiveType === 'Timestamp') {
+        resolved = {
+          ...resolved,
+          schema: { type: 'string', format: 'timestamp' },
+        };
+      }
 
       withResult(schemaTypeToModelType(name, resolved, fail.in(`property ${name}`)), (type) => {
         target[name] = {
@@ -135,6 +150,9 @@ export function importCloudFormationRegistryResource(options: LoadCloudFormation
       } else {
         switch (resolved.schema.type) {
           case 'string':
+            if (resolved.schema.format === 'timestamp') {
+              return { type: 'date-time' };
+            }
             return { type: 'string' };
 
           case 'array':
@@ -198,7 +216,12 @@ export function importCloudFormationRegistryResource(options: LoadCloudFormation
       db.link('usesType', res, typeDef);
       typeDefinitions.set(schema, typeDef);
 
-      recurseProperties(schema, typeDef.properties, fail.in(`typedef ${nameHint}`));
+      recurseProperties(
+        schema,
+        typeDef.properties,
+        fail.in(`typedef ${nameHint}`),
+        options.resourceSpec?.types?.[typeDef.name],
+      );
     }
 
     return { type: 'ref', reference: ref(typeDef) };
@@ -232,7 +255,7 @@ export function importCloudFormationRegistryResource(options: LoadCloudFormation
     const attributeNames = Array.from(
       new Set([
         ...findAttributes(source).map(simplePropNameFromJsonPtr),
-        ...Object.keys(options.specResource?.Attributes ?? {}),
+        ...Object.keys(options.resourceSpec?.spec?.Attributes ?? {}),
       ]),
     );
 

--- a/packages/@aws-cdk/service-spec-build/test/property-conversion.test.ts
+++ b/packages/@aws-cdk/service-spec-build/test/property-conversion.test.ts
@@ -1,4 +1,4 @@
-import { emptyDatabase } from '@aws-cdk/service-spec';
+import { DefinitionReference, emptyDatabase } from '@aws-cdk/service-spec';
 import { Failures } from '@cdklabs/tskb';
 import { importCloudFormationRegistryResource } from '../src/import-cloudformation-registry';
 
@@ -121,9 +121,11 @@ test('include legacy attributes in attributes', () => {
       },
       readOnlyProperties: ['/properties/Id'],
     },
-    specResource: {
-      Attributes: {
-        Property: { PrimitiveType: 'String' },
+    resourceSpec: {
+      spec: {
+        Attributes: {
+          Property: { PrimitiveType: 'String' },
+        },
       },
     },
   });
@@ -174,4 +176,48 @@ test('reference types are correctly named', () => {
 
   expect(types.length).toBe(1);
   expect(types[0].entity.name).toBe('Property');
+});
+
+test('legacy timestamps are getting the timestamp format', () => {
+  importCloudFormationRegistryResource({
+    db,
+    fails,
+    resource: {
+      description: 'Test resource',
+      typeName: 'AWS::Some::Type',
+      properties: {
+        Property: { $ref: '#/definitions/Property' },
+        Id: { type: 'string' },
+      },
+      definitions: {
+        Property: {
+          type: 'object',
+          properties: {
+            DateTime: {
+              type: 'string',
+            },
+          },
+        },
+      },
+      readOnlyProperties: ['/properties/Id'],
+    },
+    resourceSpec: {
+      spec: {},
+      types: {
+        Property: {
+          Properties: {
+            DateTime: {
+              PrimitiveType: 'Timestamp',
+              UpdateType: 'Mutable',
+            },
+          },
+        },
+      },
+    },
+  });
+
+  const prop = db.lookup('resource', 'cloudFormationType', 'equals', 'AWS::Some::Type')[0]?.properties?.Property;
+  expect(prop.type.type).toBe('ref');
+  const type = db.get('typeDefinition', (prop.type as DefinitionReference).reference.$ref);
+  expect(type.properties.DateTime.type).toMatchObject({ type: 'date-time' });
 });

--- a/packages/@aws-cdk/service-spec/src/types/resource.ts
+++ b/packages/@aws-cdk/service-spec/src/types/resource.ts
@@ -163,7 +163,7 @@ export type PropertyType =
   | MapType<PropertyType>
   | TypeUnion<PropertyType>;
 
-export type PrimitiveType = StringType | NumberType | BooleanType | JsonType | NullType;
+export type PrimitiveType = StringType | NumberType | BooleanType | JsonType | DateTimeType | NullType;
 
 export function isPrimitiveType(x: PropertyType): x is PrimitiveType {
   return (x as any).type;
@@ -192,6 +192,10 @@ export interface JsonType {
 
 export interface NullType {
   readonly type: 'null';
+}
+
+export interface DateTimeType {
+  readonly type: 'date-time';
 }
 
 export interface DefinitionReference {

--- a/packages/@cdklabs/typewriter/src/renderer/typescript.ts
+++ b/packages/@cdklabs/typewriter/src/renderer/typescript.ts
@@ -49,7 +49,7 @@ import {
 } from '../statements';
 import { StructType } from '../struct';
 import { ThingSymbol } from '../symbol';
-import { Type } from '../type';
+import { PrimitiveType, Type } from '../type';
 import { TypeParameterSpec } from '../type-declaration';
 import { Initializer, MemberVisibility, Method } from '../type-member';
 
@@ -344,7 +344,12 @@ export class TypeScriptRenderer extends Renderer {
       return this.emit('void');
     }
     if (ref.primitive) {
-      return this.emit(ref.primitive);
+      switch (ref.primitive) {
+        case PrimitiveType.DateTime:
+          return this.emit('Date');
+        default:
+          return this.emit(ref.primitive);
+      }
     }
 
     if (ref.symbol) {

--- a/packages/@cdklabs/typewriter/src/type.ts
+++ b/packages/@cdklabs/typewriter/src/type.ts
@@ -8,9 +8,9 @@ import { MemberVisibility } from './type-member';
 
 export enum PrimitiveType {
   /**
-   * A JSON date (represented as it's ISO-8601 string form).
+   * A JSON date-time or date (represented as it's ISO-8601 string form).
    */
-  Date = 'date',
+  DateTime = 'date-time',
   /**
    * A plain string.
    */
@@ -56,6 +56,7 @@ export type TypeReferenceSpec =
 export class Type {
   public static readonly ANY = new Type({ primitive: PrimitiveType.Any });
   public static readonly VOID = new Type({ primitive: PrimitiveType.Void });
+  public static readonly DATE_TIME = new Type({ primitive: PrimitiveType.DateTime });
   public static readonly STRING = new Type({ primitive: PrimitiveType.String });
   public static readonly NUMBER = new Type({ primitive: PrimitiveType.Number });
   public static readonly BOOLEAN = new Type({ primitive: PrimitiveType.Boolean });

